theano/tensor/alt_gemm_template.c

```diff
@@ -0,0 +1,133 @@
+        while(count) {
+            *((%(float_type)s*)data) *= scalar;
+            data += stride;
+            --count;
+        }
+    } while(get_next(iterator));
+    NpyIter_Deallocate(iterator);
+}
+/*Matrix+Matrix function. Compute (coeffA * matrixA) + (coeffB * matrixB)
+ * Remark: This function actually sums a C-contiguous matrix (alpha*op(A)*op(B)) with a F-contiguous matrix (beta*C)
+ * (see gemm implementation at next function for more details) */
+void alt_numpy_matrix_extended_sum_in_place_%(float_type)s(
+        const %(float_type)s* ALPHA, PyArrayObject* A,
+        const %(float_type)s* BETA, PyArrayObject* B
+) {
+    PyArrayObject* op[2]       = {A, B};
+    npy_uint32     op_flags[2] = {NPY_ITER_READONLY, NPY_ITER_READWRITE};
+    npy_uint32     flags       = 0;
+    NpyIter*       iterators   = NpyIter_MultiNew(
+            2, op, flags, NPY_CORDER, NPY_NO_CASTING, op_flags, NULL);
+    if(iterators == NULL)
+        alt_fatal_error("Unable to iterate over some matrices "
+                        "for matrix + matrix operation.");
+    NpyIter_IterNextFunc* get_next = NpyIter_GetIterNext(iterators, NULL);
+    char** data_ptr_array = NpyIter_GetDataPtrArray(iterators);
+    do {
+        %(float_type)s* from_A = (%(float_type)s*)data_ptr_array[0];
+        %(float_type)s* from_B = (%(float_type)s*)data_ptr_array[1];
+        *from_B = (*ALPHA)*(*from_A) + (*BETA)*(*from_B);
+    } while(get_next(iterators));
+    NpyIter_Deallocate(iterators);
+}
+PyObject* alt_op_without_copy_%(float_type)s(int transposable, %(float_type)s* M, int nrow, int ncol, int LDM) {
+    // By default, M is considered as a nrow*ncol F-contiguous matrix with LDM as stride indicator for the columns.
+    npy_intp dims[2];
+    npy_intp strides[2];
+    int flags;
+    if(transposable) {
+        dims[0] = ncol;
+        dims[1] = nrow;
+        strides[0] = dims[1] * %(float_size)d;
+        strides[1] = %(float_size)d;
+        flags = NPY_ARRAY_C_CONTIGUOUS;
+    } else {
+        dims[0] = nrow;
+        dims[1] = ncol;
+        strides[0] = %(float_size)d;
+        strides[1] = LDM * %(float_size)d;
+        flags = NPY_ARRAY_F_CONTIGUOUS;
+    }
+    return PyArray_New(&PyArray_Type, 2, dims, %(npy_float)s, strides, M, 0, flags, NULL);
+}
+/* %(name)s template code */
+void %(name)s(
+    char* TRANSA, char* TRANSB, 
+    const int* M, const int* N, const int* K,
+    const %(float_type)s* ALPHA, %(float_type)s* A, const int* LDA, 
+    %(float_type)s* B, const int* LDB, const %(float_type)s* BETA, 
+    %(float_type)s* C, const int* LDC) {
+    /* NB: it seems that matrix+matrix and scalar*matrix functions
+     * defined above do not allocate iterator for a matrix with 0
+     * content, that is a matrix whose nrow*ncol == 0. As these
+     * functions actually work with M*N matrices (op(A)*op(B) and/or C),
+     * I think that we could just return if M or N is null. */
+    if(*M < 1 || *N < 1 || *K < 0 || *LDA < 0 || *LDB < 0 || *LDC < 0)
```

Comment:
I think that if any of these is < 0, we should exit with an error (`alt_fatal_error`).